pyiron_continuum/fenics/wrappers.py

```diff
@@ -0,0 +1,333 @@
+# coding: utf-8
+# Copyright (c) Max-Planck-Institut für Eisenforschung GmbH - Computational Materials Design (CM) Department
+# Distributed under the terms of "New BSD License", see the LICENSE file.
+
+"""
+Classes for matching fenics objects to serializable pyiron objects.
+"""
+
+import re
+from abc import ABC, abstractmethod
+import fenics as FEN
+from fenics import (
+    near, Constant, Expression,
+    Point, BoxMesh, RectangleMesh, UnitSquareMesh, UnitCubeMesh, UnitDiscMesh, UnitTriangleMesh, UnitIntervalMesh,
+    dot, inner, grad, nabla_grad,
+)
+from mshr import (
+    generate_mesh, Tetrahedron, Box, Rectangle, Circle
+)
+from dolfin.cpp import mesh as FenicsMesh
+from pyiron_base import HasStorage
+
+
+class StringInputParser:
+    """
+    Not fool proof, but does its best to ensure that a string is valid and will execute in the scope of `known_elements`
+    (provided by the developer at instantiation) and `**kwargs` (provided by the user at call).
+    """
+    def __init__(self, known_elements):
+        self._splitting_elements = [
+            r"\*?\*", r"\+", r"-", r"/",
+            r"\(", r"\)", r",", r"\[", r"\]",
+            r"<", r">", r"=", r"==",
+            r"and", r"or", r"is", r"not",
+            r"'", r'"'
+        ]
+        self._known_elements = known_elements
+
+    def __call__(self, input_string: str, __verbose=True, **kwargs):
+        self._test_kwargs(kwargs, __verbose)
+        self._test_elements(input_string, kwargs, __verbose)
+
+    @property
+    def known_elements(self):
+        return [k.__name__ if hasattr(k, '__name__') else k for k in self._known_elements]
+
+    def _test_kwargs(self, kwargs, __verbose):
+        failures = {}
+        for key, value in kwargs.items():
+            try:
+                float(value)
+            except:
+                try:  # Recursively check if the value is an allowable type for this parser
+                    self(value, __verbose=False)
+                except:
+                    failures[key] = value
+        if failures and __verbose:
+            raise ValueError(f"Got an unexpected kwarg(s) '{failures}'")
+
+    def _split_input(self, input_string):
+        scientific_notation_re = r"[0-9]*\.[0-9]+[eE][+-]?[0-9]+|[0-9]+[eE][+-]?[0-9]+"
+        return [
+            r.strip()
+            for r in re.split(
+                "|".join(self._splitting_elements),
+                "".join(re.split(scientific_notation_re, input_string)),
+            )
+            if len(r.strip()) > 0
+        ]
+
+    def _test_elements(self, input_string, kwargs, __verbose):
+        failures = []
+        for e in self._split_input(input_string):
+            if e in self.known_elements:
+                continue
+            elif e.isnumeric():
```

Comment:
Isn't this line covered by `float(e)`?

Reply:
Yeah, and at the same time I think I'll go in and implement the `replace+isdigit` solution [from SO here](https://stackoverflow.com/questions/354038/how-do-i-check-if-a-string-is-a-number-float), as someone helpfully did profiling and shows it's fastest.

Reply:
Ok, the fancy SO answer requires it to be string-like to start with, so we'd need extra tries anyhow. I just removed the superfluous `e.isnumeric()` check.